### PR TITLE
oss-fuzz: fuzz lxc_config_define_add and lxc_config_define_load

### PR DESCRIPTION
**Pull request by evverx**
Signed-off-by: Evgeny Vereshchagin <evvers@ya.ru>

### CODE REVIEW

File: src/tests/fuzz-lxc-define-load.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	lxc_list_init(&defines);
+
+	new_str = (char *)malloc(size+1);
+	lxc_test_assert_abort(new_str);
+
+	memcpy(new_str, data, size);
+	new_str[size] = '\0';
+
+	if (lxc_config_define_add(&defines, new_str) >= 0)
+		lxc_config_define_load(&defines, c);
```
Comment (evverx) [inline src/tests/fuzz-lxc-define-load.c:31]: @brauner looks like `lxc.namespace.share=` somehow triggers
```
2021-04-02T06:05:34.6809418Z ==14==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x603000118ca4 at pc 0x0000004e5962 bp 0x7ffd1be71970 sp 0x7ffd1be71118
2021-04-02T06:05:34.6810219Z READ of size 1 at 0x603000118ca4 thread T0
2021-04-02T06:05:34.6810999Z SCARINESS: 12 (1-byte-read-heap-buffer-overflow)
2021-04-02T06:05:34.6812159Z     #0 0x4e5961 in strcmp /src/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:450:5
2021-04-02T06:05:34.6812986Z     #1 0x5d92f9 in strequal /src/lxc/src/lxc/./string_utils.h:132:9
2021-04-02T06:05:34.6813608Z     #2 0x5d9128 in lxc_namespace_2_ns_idx /src/lxc/src/lxc/namespace.c:66:7
2021-04-02T06:05:34.6814245Z     #3 0x4313a5 in set_config_namespace_share /src/lxc/src/lxc/confile.c:2918:11
2021-04-02T06:05:34.6815041Z     #4 0x5ad26c in lxc_set_config_item_locked /src/lxc/src/lxc/lxccontainer.c:3144:8
2021-04-02T06:05:34.6815745Z     #5 0x5bdc96 in do_set_config_item_locked /src/lxc/src/lxc/lxccontainer.c:3169:8
2021-04-02T06:05:34.6816448Z     #6 0x5bdc04 in do_lxcapi_set_config_item /src/lxc/src/lxc/lxccontainer.c:3186:6
2021-04-02T06:05:34.6817151Z     #7 0x5b0dc3 in lxcapi_set_config_item /src/lxc/src/lxc/lxccontainer.c:3192:1
2021-04-02T06:05:34.6817795Z     #8 0x583ca5 in lxc_config_define_load /src/lxc/src/lxc/confile.c:3106:10
2021-04-02T06:05:34.6818800Z     #9 0x581125 in LLVMFuzzerTestOneInput /src/lxc/src/tests/fuzz-lxc-define-load.c:31:3
2021-04-02T06:05:34.6822207Z     #10 0x483b93 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
2021-04-02T06:05:34.6824159Z     #11 0x48309a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
2021-04-02T06:05:34.6825755Z     #12 0x484df7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
2021-04-02T06:05:34.6827580Z     #13 0x485985 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
2021-04-02T06:05:34.6829369Z     #14 0x473717 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
2021-04-02T06:05:34.6830580Z     #15 0x49f572 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
2021-04-02T06:05:34.6831549Z     #16 0x7f4cad3ce83f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
2021-04-02T06:05:34.6832454Z     #17 0x448688 in _start (out/fuzz-lxc-define-load+0x448688)
2021-04-02T06:05:34.6832905Z 
2021-04-02T06:05:34.6833530Z DEDUP_TOKEN: strcmp--strequal--lxc_namespace_2_ns_idx
2021-04-02T06:05:34.6834536Z 0x603000118ca4 is located 0 bytes to the right of 20-byte region [0x603000118c90,0x603000118ca4)
2021-04-02T06:05:34.6835100Z allocated by thread T0 here:
2021-04-02T06:05:34.6835951Z     #0 0x5391f4 in __strdup /src/llvm-project/compiler-rt/lib/asan/asan_interceptors.cpp:468:3
2021-04-02T06:05:34.6836688Z     #1 0x5837be in parse_new_conf_line /src/lxc/src/lxc/confile.c:3052:6
2021-04-02T06:05:34.6837304Z     #2 0x582f1a in lxc_config_define_add /src/lxc/src/lxc/confile.c:3090:15
2021-04-02T06:05:34.6838301Z     #3 0x5810fa in LLVMFuzzerTestOneInput /src/lxc/src/tests/fuzz-lxc-define-load.c:30:6
2021-04-02T06:05:34.6839813Z     #4 0x483b93 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
2021-04-02T06:05:34.6841530Z     #5 0x48309a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
2021-04-02T06:05:34.6843085Z     #6 0x484df7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
2021-04-02T06:05:34.6844755Z     #7 0x485985 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
2021-04-02T06:05:34.6846485Z     #8 0x473717 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
2021-04-02T06:05:34.6847707Z     #9 0x49f572 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
2021-04-02T06:05:34.6848646Z     #10 0x7f4cad3ce83f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
2021-04-02T06:05:34.6849076Z 
2021-04-02T06:05:34.6849701Z DEDUP_TOKEN: __strdup--parse_new_conf_line--lxc_config_define_add
2021-04-02T06:05:34.6851086Z SUMMARY: AddressSanitizer: heap-buffer-overflow /src/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:450:5 in strcmp
2021-04-02T06:05:34.6852119Z Shadow bytes around the buggy address:
2021-04-02T06:05:34.6852602Z   0x0c068001b140: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
2021-04-02T06:05:34.6853109Z   0x0c068001b150: fa fa fa fa fa fa fd fd fd fd fa fa fd fd fd fa
2021-04-02T06:05:34.6853593Z   0x0c068001b160: fa fa fa fa fa fa fa fa fd fd fd fa fa fa fa fa
2021-04-02T06:05:34.6854077Z   0x0c068001b170: fa fa fa fa fa fa fa fa fa fa fd fd fd fa fa fa
2021-04-02T06:05:34.6854701Z   0x0c068001b180: fa fa fa fa fa fa fa fa fa fa fa fa fd fd fd fa
2021-04-02T06:05:34.6855205Z =>0x0c068001b190: fa fa 00 00[04]fa fa fa fa fa fa fa fa fa fd fd
2021-04-02T06:05:34.6855693Z   0x0c068001b1a0: fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa
2021-04-02T06:05:34.6856184Z   0x0c068001b1b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
2021-04-02T06:05:34.6856670Z   0x0c068001b1c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fd fd
2021-04-02T06:05:34.6857181Z   0x0c068001b1d0: fd fa fa fa fd fd fd fd fa fa fa fa fa fa fa fa
2021-04-02T06:05:34.6857670Z   0x0c068001b1e0: fa fa fa fa fa fa fd fd fd fd fa fa fd fd fd fd
2021-04-02T06:05:34.6858369Z Shadow byte legend (one shadow byte represents 8 application bytes):
2021-04-02T06:05:34.6858932Z   Addressable:           00
2021-04-02T06:05:34.6859355Z   Partially addressable: 01 02 03 04 05 06 07 
2021-04-02T06:05:34.6859803Z   Heap left redzone:       fa
2021-04-02T06:05:34.6860194Z   Freed heap region:       fd
2021-04-02T06:05:34.6860584Z   Stack left redzone:      f1
2021-04-02T06:05:34.6860981Z   Stack mid redzone:       f2
2021-04-02T06:05:34.6861361Z   Stack right redzone:     f3
2021-04-02T06:05:34.6861749Z   Stack after return:      f5
2021-04-02T06:05:34.6862150Z   Stack use after scope:   f8
2021-04-02T06:05:34.6862528Z   Global redzone:          f9
2021-04-02T06:05:34.6862904Z   Global init order:       f6
2021-04-02T06:05:34.6863287Z   Poisoned by user:        f7
2021-04-02T06:05:34.6863737Z   Container overflow:      fc
2021-04-02T06:05:34.6864120Z   Array cookie:            ac
2021-04-02T06:05:34.6864509Z   Intra object redzone:    bb
2021-04-02T06:05:34.6864894Z   ASan internal:           fe
2021-04-02T06:05:34.6865285Z   Left alloca redzone:     ca
2021-04-02T06:05:34.6865689Z   Right alloca redzone:    cb
2021-04-02T06:05:34.6866053Z   Shadow gap:              cc
2021-04-02T06:05:34.6866375Z ==14==ABORTING
2021-04-02T06:05:34.6867896Z MS: 5 CopyPart-EraseBytes-ChangeByte-InsertByte-CMP- DE: "lxc.namespace.share"-; base unit: e205c79e481d7a9cbd85e5eb3d3520c9cd136031
2021-04-02T06:05:34.6870865Z 
```
Could you take a look?

Comment (brauner) [reply src/tests/fuzz-lxc-define-load.c:31]: I think this is the fix:
https://github.com/lxc/lxc/pull/3764

Feel free to pick this into your branch here.

**Comment by brauner** [comment]
jenkins: test this please

File: src/tests/fuzz-lxc-define-load.c (hunk 1 above)
Comment (evverx) [reply src/tests/fuzz-lxc-define-load.c:31]: It seems the fuzzer no longer crashes. Thanks!

File: src/lxc/confile.c [hunk 2]
```diff
@@ -3111,7 +3111,9 @@ bool lxc_config_define_load(struct lxc_list *defines, struct lxc_container *c)
 			break;
 	}
 
+#ifndef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
 	lxc_config_define_free(defines);
+#endif /* !FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION */
```
Comment (evverx) [inline src/lxc/confile.c:3116]: I have to say I'm not sure I like this kludge :-) but it seems to be the easiest way to keep all the keys `lxc_config_define_load` has just processed to make it possible to iterate over them to make sure `get_config_item` can properly extract their values (previously set with `set_config_item`). Another option would be to copy the keys somewhere between lxc_config_define_add and lxc_config_define_load but it seems to slow down the fuzzer unnecessarily. @brauner I'd appreciate it if you could take a look and point me in the right direction :-)

Comment (evverx) [reply src/lxc/confile.c:3116]: The idea is to cover the "get_*" functions from `config_jump_table`. It should help to trigger scenarios where `set_config_*` followed by `get_config_*` can cause unexpected side effects like half-initialiazied structures and so on

Comment (brauner) [reply src/lxc/confile.c:3116]: Hm, I think for now that your hack is fine. :)

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
Thanks!

**Comment by brauner** [comment]
I think one of the next things we should think about tackling is `cgroup.c` and the various callbacks associated with `struct cgroup_ops`.

**Pull request merged**